fineract-provider/src/main/java/org/apache/fineract/portfolio/note/domain/NoteFactory.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.fineract.portfolio.note.domain;
+
+import org.apache.fineract.portfolio.client.domain.Client;
+import org.apache.fineract.portfolio.group.domain.Group;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransaction;
+import org.apache.fineract.portfolio.savings.domain.SavingsAccount;
+import org.apache.fineract.portfolio.savings.domain.SavingsAccountTransaction;
+import org.apache.fineract.portfolio.shareaccounts.domain.ShareAccount;
+
+public final class NoteFactory {
```

Comment:
hi @Zeyad2003, I came across your PR today. I would suggest either using the builder, or using the factory, but putting a factory layer facading a builder layer seems to be an overkill to me.

Please give it a try to refactor going with either of the approaches and see if the code gets simpler and more straightforward.. Thanks!

Reply:
Hi @kjozsa, Thanks for your feedback.

I understand your point, and I'd agree but in different scenarios that don't involve JPA entities.

Using public `@Builder` may cause some violations like bypassing some validation checks, or improperly updating fields that should remain immutable after instantiation. 
> That's why I limited the access to `AccessLevel.PACKAGE`

Relying solely on *Factory*  would push us towards less flexible solutions like:
- Gigantic constructor with +30 parameter or multiple parameterized constructors definition (not flexible).

- Using some static methods that have their own problems but also violate the Single Responsibility Principle (SRP) and at the end of the day we might face the same constructor problem as well.

You can review the earlier discussion on using `@Builder` with JPA entities [here](https://lists.apache.org/thread/l2xdp07984bl4x1h34hyymr9q3o6qxxo).

Please let me know your thoughts.

Thank you!
Zeyad

Reply:
Ok, I agree with your reasons, but in that case, I would enforce everyone to use the dedicated Factory for Note instantiation, so we'd need to get rid of the public constructors. 

I would also consider moving the `Note#update(JsonCommand)` function's functionality away from the entity, as this makes a direct dependency between the entity layer and the web layer..

Reply:
Done.
It has been replaced by a private `update()` method within the `NoteWritePlatformService`.